### DISCUSSION TITLE
Hooks work well in nestjs-typegoose

### DISCUSSION THREAD

**Discussion by smolinari** [GrapeoffJS/kindagoose#116] [General]
Hi,

What do you mean with this in the docs? 

> But as it turned out, on nestjs-typegoose there are no comfortable instruments to work with hooks. 

Hooks work just fine in Typegoose and nestjs-typegoose has no issues with them either. 

https://typegoose.github.io/typegoose/docs/api/decorators/hooks

Scott

**Comment by findingtheconcept** [comment]
Hello, @smolinari! 

Hooks are great in pure node.js applications, but in nest.js the are completely useless without dependency injection (at least for me). So, the motivation of creating the `kindagoose` was allowing users to use dependency injection in hooks.

https://grapeoffjs.github.io/kindagoose/#/pages/event-tracking

**Reply by smolinari** [reply]
Ah. Ok. So, if that is the selling point, you should explain it that way too, instead of saying something that isn't true.  

Scott

**Reply by findingtheconcept** [reply]
The nestjs-typegoose actually does not have convenient tools for dealing with hooks. Of course, pure typegoose does, but the main purpose of the first one is to make full connection of typegoose and nestjs.

So, I think in that way my statement looks truly, isn't it?

**Reply by findingtheconcept** [reply]
You are welcome to suggest a correction for docs, this is an open source.

**Reply by smolinari** [reply]
nestjs-typegoose is just a Nestjs module wrapper for Typegoose. So, the user can use Typegoose as it is with Nest. It is "pure Typegoose" and hooks work exactly the same. 

Btw, needing DI inside Mongoose hooks is an anti-pattern to me, because the things you should be doing with hooks should only be on the injected document or query in the hook. If you are doing side-effects like hitting other services, that kind of logic is something that should be happening in your services before or after using the model. 

Also, your example of needing to feed elasticsearch would be much better off inside a Mongoose plugin, much like Mongoosastic.  https://github.com/mongoosastic/mongoosastic

Using a Mongoose plugin hides the complexity of needing to write hooks on all your entities to index in ES. You just add the plugin to your entity or even globally, if all entities should be getting indexed. The only thing you'd have to do after that is flag which fields you want indexed, which the plugin adds to the `@Prop` options. 

```ts
class User {
  @prop({es_index: true}) // <- like here
  public name?: string;

  @prop({ required: true })
  public age!: number; // This is a single Primitive

  @prop({ type: () => [String] })
  public preferences?: string[]; // This is a Primitive Array
...
}
``` 
See the docs on all the options offered too. Not bad, if you ask me: https://mongoosastic.github.io/mongoosastic/index.html

Unfortunately, Mogoosastic is lagging in maintenance. I guess the need for proper and much more powerful text indexing isn't a big requirement with most Mongodb users. Still, a plugin solution for feeding ES is much, much better than using hooks. :grin: 

Scott

**Reply by findingtheconcept** [reply]
The last time I used mongoosastic it did not work for me and I don't know why. I had to write my own implementation of indexing the documents and the first problem for me was segregating CRUD and elastic.

I had to change the index after every single mutation of a mongodb document.

The first attempt was to index them manually after every mutating repository method call by calling the elasticService. 

And I have thought that it will be cool to delegate this task to some kind of event listeners, to make it more clear. Instead of calling a needed method of elasticSearch Manually, I just should call it once in an event handler. But the first requirement was to make with the minimal changes in code.

So, without supporting the dependency injection for that in nestjs-typegoose I needed to make singleton instance of ElasticSearch outside of DI. (I found the solution later, but nod as good as using DI in hooks)

**Reply by smolinari** [reply]
Ok. Well, I guess to learn this isn't a bad direction to take. But reinventing wheels, because the other wheels don't work as expected is a lot of additional work than trying to get the old wheels to work right. :grin: 

Keep on trying tho.... it's not a bad thing at all. 

Scott

**Comment by findingtheconcept** [comment]
You are always welcome to use Typegoose's native form of registering hooks, I only offer you a more «Nesty» way.
